### ISSUE TITLE
xml osm data import errors

### ISSUE THREAD

**Issue by sethvincent** [osmlab/field-data-collection#168]
```
I/ReactNativeJS( 3447): syncSurveyData
I/ReactNativeJS( 3447): 'checkOsmMeta shouldSync?', true
I/ReactNativeJS( 3447): replicate
I/ReactNativeJS( 3447): db.clear: 1
I/ReactNativeJS( 3447): 'db.clear: 2; pending ===', 3, 'type ===', 'db'
I/ReactNativeJS( 3447): 'db.clear: 2; pending ===', 2, 'type ===', 'index'
I/ReactNativeJS( 3447): 'db.clear: 2; pending ===', 1, 'type ===', 'chunk-store'
I/ReactNativeJS( 3447): start replicateOsmOrgDb
E/ReactNativeJS( 3447): Storage is closed
W/ReactNativeJS( 3447): Unable to symbolicate stack trace: The stack is null
I/ReactNativeJS( 3447): 'finished replicating', 2
I/ReactNativeJS( 3447): netSync.replicate finished
I/ReactNativeJS( 3447): closed and reopened osmorgdb
I/ReactNativeJS( 3447): osm.replicate finished
E/ReactNativeJS( 3447): undefined is not an object (evaluating 'this.fun.apply')
W/ReactNativeJS( 3447): Unable to symbolicate stack trace: The stack is null
E/ReactNativeJS( 3447): corrupt data: region cycle detected
W/ReactNativeJS( 3447): Unable to symbolicate stack trace: The stack is null
E/ReactNativeJS( 3447): corrupt data: region cycle detected
W/ReactNativeJS( 3447): Unable to symbolicate stack trace: The stack is null
```

@noffle Things seem to go wrong initially because the chunk store is returning as closed when `replicateOsmOrgDb` starts.

**Comment by sethvincent** [comment]
Made this branch with some of the debugging attempts I was trying out: https://github.com/osmlab/field-data-collection/tree/debug-xml-import

**Comment by noffle** [comment]
I pushed to your branch to add some more debug output that could be helpful for seeing what's going on with the chunk store: 57f8f7c

**Comment by sethvincent** [comment]
Woops i left that `destroyAllData` action a little messier than I thought.

Without using it and starting from a fresh install of the app output looks slightly different, but still has roughly the same errors:

```
I/ReactNativeJS( 3687): syncSurveyData
I/ReactNativeJS( 3687): 'checkOsmMeta shouldSync?', true
I/ReactNativeJS( 3687): replicate
I/ReactNativeJS( 3687): '0', '38520829565823'
I/ReactNativeJS( 3687): '1', '38520829565823'
I/ReactNativeJS( 3687): db.clear: 1
I/ReactNativeJS( 3687): 'db.clear: 2; pending ===', 3, 'type ===', 'index'
I/ReactNativeJS( 3687): 'db.clear: 2; pending ===', 2, 'type ===', 'db'
I/ReactNativeJS( 3687): 'db.clear: 2; pending ===', 1, 'type ===', 'chunk-store'
I/ReactNativeJS( 3687): '2', '6955199036747'
I/ReactNativeJS( 3687): '3', '6955199036747'
I/ReactNativeJS( 3687): start replicateOsmOrgDb
I/ReactNativeJS( 3687): 'finished replicating', 2
I/ReactNativeJS( 3687): '4', '6955199036747'
I/ReactNativeJS( 3687): netSync.replicate finished
I/ReactNativeJS( 3687): closed and reopened osmorgdb
I/ReactNativeJS( 3687): '5', '6955199036747'
I/ReactNativeJS( 3687): '6', '77281822264194'
I/ReactNativeJS( 3687): osm.replicate finished
E/ReactNativeJS( 3687): undefined is not an object (evaluating 'this.fun.apply')
W/ReactNativeJS( 3687): Unable to symbolicate stack trace: The stack is null
E/ReactNativeJS( 3687): 'Unhandled promise rejection', { [Error: write callback called multiple times]
E/ReactNativeJS( 3687):   line: 97557,
E/ReactNativeJS( 3687):   column: 9,
E/ReactNativeJS( 3687):   sourceURL: 'http://10.0.2.2:8081/index.android.bundle?platform=android&dev=true&hot=false&minify=false' }
E/ReactNativeJS( 3687): Storage is closed
W/ReactNativeJS( 3687): Unable to symbolicate stack trace: The stack is null
W/ReactNativeJS( 3687): Unable to symbolicate stack trace: The stack is null
E/ReactNativeJS( 3687): Database is not open
```

**Comment by noffle** [comment]
@sethvincent Where is the new `Storage is closed` being called from? I wonder if we're leaking a reference to `osmOrgDb` to another module, and it's holding onto the old (closed) DB.

**Comment by sethvincent** [comment]
It's the `get` method of asyncstorage-chunk-store.

**Comment by noffle** [comment]
Do you have a stack trace?

**Comment by sethvincent** [comment]
I wish. So far I'm pretty sure that the issue is related to the `closeAndReopenOsmOrgDb` function. Or at least, if I don't call it the errors don't show up.

What if it's something to do with timing? Like the observations db is trying to do work while the store is closed. That's my current hypothesis.

**Comment by noffle** [comment]
I pushed b5c44d0, which waits on *all* 3 backends (hyperlog, index, chunk-store). I think was this hanging before (chunk store not closing?), but maybe that's not the case with AsyncStorage now, since you made a number of fixes to it since then.

**Comment by noffle** [comment]
BTW @sethvincent, thanks for being so patient with this sorta tedious debugging back-and-forth. :heart:

**Comment by sethvincent** [comment]
Thanks.

It looks like indexing takes quite a while.

As an experiment I made `OsmSync.replicate` only replicate the osm db, not the observations. Everything worked as expected. It took about 7 minutes for indexing to complete, but it didn't hit the errors I saw above.

Any ideas based on that? Doesn't make a lot of sense.

**Comment by sethvincent** [comment]
Just pushed some more changes with a couple interesting bits:

**Comment by sethvincent** [comment]
fixed by #176
